### ISSUE TITLE
`EloquentCollection::map` callback expects Model instead of concrete model type

### ISSUE THREAD

**Issue by alies-dev** [psalm/psalm-plugin-laravel#514]
## Description

When calling methods on an Eloquent Relation (e.g., `findOrFail()`, `first()`, `where()`), the concrete model type from the relation's generic parameter is lost. The template parameter `TRelatedModel` resolves to the constraint `Model` instead of the concrete type (e.g., `File`).

This causes `InvalidPropertyAssignmentValue` (86 occurrences) and `ArgumentTypeCoercion` on `Collection::map` callbacks (35 occurrences) — **121 false positives total** from this single root cause.

## Reproduction

Found while auditing [monicahq/monica](https://github.com/monicahq/monica) with plugin v4.2.0 + Psalm 7.0.0-beta17.

## Root Cause Analysis

Traced the type resolution step by step using `@psalm-trace`:

```php
/**
 * @param HasMany<File, Vault> $relation
 */
public function test(HasMany $relation): void
{
    $relation;              // HasMany<File, Vault>              ✅
    $relation->where(...);  // Builder<TRelatedModel as Model>   ❌ loses File
    $relation->findOrFail(1); // TRelatedModel as Model          ❌ loses File
    $relation->first();     // TRelatedModel as Model | null     ❌ loses File
    $relation->get();       // Collection<int, File>             ✅ works!
}
```

### Why `get()` works but others don't

- **`get()`** is declared directly on `Relation` with its own template params — Psalm resolves them correctly.
- **`where()`, `findOrFail()`, `first()`** are forwarded via `Relation::__call()` → `Builder`. The `RelationsMethodHandler` intercepts these calls, but `$event->getTemplateTypeParameters()` already contains the unresolved `TRelatedModel` bound to the constraint `Model` rather than the concrete `File`.

### Where exactly the type is lost

The `RelationsMethodHandler` (line 70-83) checks if a Builder method returns `$this`/`static`:
- If yes (e.g., `where`): returns the Relation type with template params → but the params are already `<Model, Model>` instead of `<File, Vault>`
- If no (e.g., `findOrFail`, `first`): returns `null`, letting Psalm resolve via `__call` → returns `mixed` or falls back to the template constraint `Model`

The plugin can't fix this in `RelationsMethodHandler` because Psalm's event already provides the wrong template params. This appears to be a **Psalm core limitation** in how template type parameters are propagated through `__call` magic method forwarding.

### Additional finding: docblock generics dropped on method call

Even when the relation method has a proper docblock:
```php
// Vault model
/**
 * @return HasMany<File, $this>
 */
public function files(): HasMany { ... }
```

Calling `$vault->files()` returns `HasMany<Model, Model>` instead of `HasMany<File, Vault>`. The docblock generic params are being dropped.

## Impact

This is the **single largest source of false positives** in the Monica audit:
- `InvalidPropertyAssignmentValue`: 86 (from `$this->file = $relation->findOrFail(...)`)
- `ArgumentTypeCoercion` on `Collection::map`: 35 (from `$relation->get()->map(fn (ConcreteModel $m) => ...)`)

The pattern is universal — virtually every Laravel app chains `where()`, `first()`, `findOrFail()` on relations.

## Possible Fix Approaches

1. **Plugin-level**: Add a `MethodReturnTypeProvider` for Relation subclasses that, for non-Builder-returning methods like `findOrFail`/`first`, manually resolves the `TRelatedModel` from the relation's generic params and substitutes it in the Builder method's return type.

2. **Psalm core**: Fix template parameter propagation through `__call` so that when `HasMany<File, Vault>::__call('findOrFail', ...)` forwards to `Builder::findOrFail()`, the `TModel` param on Builder is correctly bound to `File`.

3. **Hybrid**: For the docblock-generics-dropped issue (Finding 2), investigate why `$vault->files()` loses the `@return HasMany<File, $this>` annotation. This may be a separate Psalm core issue with `$this` template resolution.
